sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/ServerCertificateCheck.kt

```diff
@@ -19,44 +19,107 @@
+import org.sonarsource.kotlin.api.ConstructorMatcher
+import org.sonarsource.kotlin.api.FunMatcher
 import org.sonarsource.kotlin.plugin.KotlinFileContext
 
+
+private const val CERTIFICATE_EXCEPTION = "CertificateException"
+
 @Rule(key = "S4830")
 class ServerCertificateCheck : AbstractCheck() {
     companion object {
-        private val methodNames = setOf("checkClientTrusted", "checkServerTrusted")
-        private val firstArgRegex = Regex("""Array<(out )?X509Certificate\??>\??""")
-        private val secondArgRegex = Regex("""String\??""")
+        val funMatchers = listOf(
+            FunMatcher {
+                definingSupertype = "javax.net.ssl.X509TrustManager"
+                withNames("checkClientTrusted", "checkServerTrusted")
+            },
+            FunMatcher {
+                definingSupertype = "javax.net.ssl.X509ExtendedTrustManager"
+                withNames("checkClientTrusted", "checkServerTrusted")
+            })
     }
 
-    override fun visitClassOrObject(node: KtClassOrObject, kotlinFileContext: KotlinFileContext) {
+    override fun visitNamedFunction(function: KtNamedFunction, kotlinFileContext: KotlinFileContext) {
         val (_, _, bindingContext) = kotlinFileContext
 
-        val extendsX509 = bindingContext.get(CLASS, node)?.getAllSuperClassifiers()?.any {
-            it.fqNameOrNull()?.asString() == "javax.net.ssl.X509TrustManager"
-        } ?: return
-        if (extendsX509) {
-            node.body?.functions?.forEach { f ->
-                if (methodNames.contains(f.name) 
-                    && f.hasCompliantParameters(bindingContext)
-                    && f.listStatements().none { it.throwsException(bindingContext) }) {
-                    kotlinFileContext.reportIssue(f.nameIdentifier ?: f,
-                        "Enable server certificate validation on this SSL/TLS connection.")
-                }
+        if (function.belongsToTrustManagerClass(bindingContext)
+            && !function.callsCheckTrusted(bindingContext)
+            && !function.throwsCertificateExceptionWithoutCatching(bindingContext)
+        ) {
+            kotlinFileContext.reportIssue(function.nameIdentifier ?: function,
+                "Enable server certificate validation on this SSL/TLS connection.")
+        }
+    }
+
+    private fun KtNamedFunction.belongsToTrustManagerClass(bindingContext: BindingContext): Boolean {
+        return funMatchers.any { it.matches(this, bindingContext) }
+    }
+
+    /*
+     * Returns true if a function contains a call to "checkClientTrusted" or "checkServerTrusted".
+     */
+    private fun KtNamedFunction.callsCheckTrusted(bindingContext: BindingContext): Boolean {
+        val visitor = object : KtVisitorVoid() {
+            private var foundCheckTrustedCall: Boolean = false
+
+            override fun visitCallExpression(expression: KtCallExpression) {
+                foundCheckTrustedCall = funMatchers.any { it.matches(expression, bindingContext) }
```

Comment:
Be careful, you still need a conditional statement here. Add this in the Sample file after line 90 to test this case:
```
                getAcceptedIssuers()
```